### PR TITLE
Fixed #13378 - Added skipOnEmpty option to SluggableBehaviour

### LINKED ISSUE #13378: SluggableBehavior and empty $attribute value
### What steps will reproduce the problem?

```php
    /**
     * @return array
     */
    public function behaviors()
    {
        $behaviors = parent::behaviors();
        $behaviors['slug'] = [
            'class' => SluggableBehavior::className(),
            'attribute' => 'title',
            'slugAttribute' => 'code',
            'ensureUnique' => true,
            'immutable' => true
        ];

        return $behaviors;
    }
```

### What is the expected result?
SluggableBehavior should not runs if model `title` attribute is empty. In example above if value of `title` attribute is not set, value of `code` attribute should not changes.

### What do you get instead?
In example above if value of `title` attribute is not set, value of `code` attribute sets to '-2'.

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.10?
| PHP version      | 7.0.13
| Operating system | Debian 8.1

### PR DESCRIPTION
**Pull request by andrewnester**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13378

### CODE REVIEW

File: tests/framework/behaviors/SluggableBehaviorTest.php [hunk 1]
```diff
@@ -229,4 +245,20 @@ public function behaviors()
             ],
         ];
     }
-}
\ No newline at end of file
+}
+
+class ActiveRecordSluggableSkipOnEmpty extends ActiveRecordSluggable
+{
```
Comment (dynasource) [inline tests/framework/behaviors/SluggableBehaviorTest.php]: the naming should have ActiveRecord at the end (ActiveRecordSluggable actually too IMO). So a better name is `SkipOnEmptySluggableActiveRecord'. Perhabs you can change that too.

**Review by dynasource — CHANGES_REQUESTED** [review]

**Comment by andrewnester** [comment]
@dynasource thanks! I've just updated my PR.

File: framework/behaviors/SluggableBehavior.php [hunk 2]
```diff
@@ -144,7 +150,9 @@ protected function getValue($event)
                 foreach ((array) $this->attribute as $attribute) {
                     $slugParts[] = ArrayHelper::getValue($this->owner, $attribute);
```
Comment (samdark) [inline framework/behaviors/SluggableBehavior.php]: Empty check and return could be done right here so there's no need to iterate array one more time.

Comment (samdark) [inline framework/behaviors/SluggableBehavior.php]:
```php
$part = ArrayHelper::getValue($this->owner, $attribute);
if ($part === null || $part === '') {
    return $this->owner->{$this->slugAttribute};
}
$slugParts[] = $part;
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/behaviors/SluggableBehavior.php (hunk 2 above)
Comment (andrewnester) [inline framework/behaviors/SluggableBehavior.php]: @samdark thanks! looks like it would be good change for performance. done!

**Review by samdark — APPROVED** [review]
Looks good. Please add a changelog entry.

File: framework/behaviors/SluggableBehavior.php [hunk 3]
```diff
@@ -142,9 +148,12 @@ protected function getValue($event)
             if ($this->isNewSlugNeeded()) {
                 $slugParts = [];
                 foreach ((array) $this->attribute as $attribute) {
-                    $slugParts[] = ArrayHelper::getValue($this->owner, $attribute);
+                    $part = ArrayHelper::getValue($this->owner, $attribute);
+                    if ($this->skipOnEmpty && $this->isEmpty($part)) {
+                        return $this->owner->{$this->slugAttribute};
```
Comment (klimov-paul) [inline framework/behaviors/SluggableBehavior.php]: I am unsure if this logic is correct:
- if there is no slug (e.g. for new record) and its source is empty - the slug will be empty.
- if slug is set (existing record) and its source is empty (but changed!) - the slug remains unchanged, which seems odd

Comment (samdark) [inline framework/behaviors/SluggableBehavior.php]: I'd expect that. Your first point is how it behaves now. Usually if you're using a field to generate a slug the field is required.  Second point is what this property was introduced for.

**Review by dynasource — APPROVED** [review]

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by rob006** [comment]
Test are broken: https://travis-ci.org/yiisoft/yii2/builds/249564948

**Pull request merged**